src/uct/ib/ud/base/ud_inl.h

```diff
@@ -79,15 +79,53 @@ uct_ud_skb_release(uct_ud_send_skb_t *skb, int is_inline)
     }
 }
 
+#if UCS_ENABLE_ASSERT
+static UCS_F_ALWAYS_INLINE int uct_ud_ep_has_pending(uct_ud_ep_t *ep)
+{
+    return !ucs_arbiter_group_is_empty(&ep->tx.pending.group) &&
+           !ucs_arbiter_elem_is_only(&ep->tx.pending.elem);
+}
+#endif
+
+static UCS_F_ALWAYS_INLINE int uct_ud_ep_has_pending_flag(uct_ud_ep_t *ep)
+{
+    return (ep->dest_ep_id & UCT_UD_EP_HAS_PENDING) != 0;
+}
+
+static UCS_F_ALWAYS_INLINE void uct_ud_ep_set_has_pending_flag(uct_ud_ep_t *ep)
+{
+    ep->dest_ep_id |= UCT_UD_EP_HAS_PENDING;
+}
+
+static UCS_F_ALWAYS_INLINE void uct_ud_ep_remove_has_pending_flag(uct_ud_ep_t *ep)
+{
+    ucs_assert(uct_ud_ep_has_pending_flag(ep) &&
+               (ep->dest_ep_id != UCT_UD_EP_NULL_ID));
+    ep->dest_ep_id &= ~UCT_UD_EP_HAS_PENDING;
+}
+
+static UCS_F_ALWAYS_INLINE void uct_ud_ep_set_dest_ep_id(uct_ud_ep_t *ep,
+                                                         uint32_t dest_id)
+{
+    ucs_assert(uct_ud_ep_has_pending_flag(ep) || !uct_ud_ep_has_pending(ep));
+    ep->dest_ep_id = dest_id | (ep->dest_ep_id & UCT_UD_EP_HAS_PENDING);
+}
+
+static UCS_F_ALWAYS_INLINE uint32_t uct_ud_ep_get_dest_ep_id(uct_ud_ep_t *ep)
+{
+    return ep->dest_ep_id & ~UCT_UD_EP_HAS_PENDING;
+}
+
 /* same as above but also check ep resources: window&connection state */
 static UCS_F_ALWAYS_INLINE uct_ud_send_skb_t *
 uct_ud_ep_get_tx_skb(uct_ud_iface_t *iface, uct_ud_ep_t *ep)
 {
-    if (ucs_unlikely(!uct_ud_ep_is_connected(ep) ||
+    if (ucs_unlikely(!uct_ud_ep_is_connected_and_no_pending(ep) ||
```

Comment:
imo, using this instead of one additional check looks overkill. 
I'd go with just adding extra condition

Reply:
@yosefe I agree with @brminich , but we need to decide if it is ok to have this complex logic or better to have one more check

Reply:
can we make HAS_PENDING part of ep->flags?
in order to unite checks, we can make is_connected a flag as well, instead of checking dest_ep_id

Reply:
yes, we can. it will be two checks, is it ok?

Reply:
@yosefe @brminich done, could you review pls?